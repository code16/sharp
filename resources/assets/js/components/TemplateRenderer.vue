<template>
    <rendered-template
        class="SharpTemplate"
        :name="name"
        :template="template"
        :template-data="templateData"
        :template-props="templateProps"
    >
    </rendered-template>
</template>

<script>
    export default {
        name: 'SharpTemplate',

        components: {
            RenderedTemplate: {
                functional: true,
                props: {
                    name:String, template:String, templateProps:Array, templateData:Object
                },
                render(h, { props, data }) {
                    const { name ,template, templateProps, templateData } = props;
                    return h({
                            name: `SharpTemplate${name}`,
                            template:`<div>${template}</div>`,
                            props: templateProps
                        }, {
                        ...data,
                        props: templateData
                    });
                }
            }
        },

        props: {
            name: String,
            templateData: Object,
            template: String,
        },

        computed: {
            templateProps() {
                return Object.keys(this.templateData||{});
            },
        }
    }
</script>