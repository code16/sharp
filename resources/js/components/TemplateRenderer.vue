<script lang="ts">
    import { Card, CardContent } from '@/components/ui/card';
    import { Badge } from '@/components/ui/badge';

    export const components = {
        'sharp-card': Card,
        'sharp-card-content': CardContent,
        'sharp-badge': Badge,
    };
</script>
<script lang="ts" setup>
    import { computed } from 'vue';

    const props = defineProps<{
        templateData?: Record<string, any>,
        templateProps?: string[],
        template: string,
    }>();

    const component = computed(() => ({
        components,
        template: `<div class="SharpTemplate">${props.template ?? ''}</div>`,
        props: [
            ...(props.templateProps || []),
            ...Object.keys(props.templateData ?? {}),
        ],
    }));
</script>

<template>
    <component :is="component" v-bind="templateData ?? {}">
        <slot />
    </component>
</template>

