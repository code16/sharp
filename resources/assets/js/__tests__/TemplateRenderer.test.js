import Vue from 'vue';
import TemplateRenderer from '../components/TemplateRenderer';



describe('template-renderer',()=>{
    Vue.component('sharp-template', TemplateRenderer);

    beforeEach(()=>{
        document.body.innerHTML = `
            <div id="app">
                <sharp-template name="Test" :template-data="{ name:'Antoine' }" :template="'<em>{{name}}</em>'"></sharp-template>
            </div>
        `
    });

    test('can mount given template wrapped in div tag', async () => {
        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });
});


async function createVm(customOptions={}) {
    const vm = new Vue({
        el: '#app',
        mixins: [customOptions],
    });

    await Vue.nextTick();

    return vm.$children[0];
}