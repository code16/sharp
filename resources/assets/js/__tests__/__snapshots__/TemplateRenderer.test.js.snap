// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template-renderer can mount given template wrapped in div tag 1`] = `
<div id="app">
	<div class="SharpTemplate">
		<em>Antoine</em>
	</div>
</div>
`;
